finality-aleph/src/block/substrate/verification/cache.rs

```diff
@@ -82,42 +77,48 @@ impl Display for CacheError {
 #[derive(Clone)]
 struct CachedData {
     session_verifier: SessionVerifier,
-    aura_authorities: Vec<(Option<AccountId>, AuraId)>,
+    aura_authorities: Vec<AuraId>,
+    authority_accounts: Option<Vec<AccountId>>,
 }
 
 fn download_data<AP: AuthorityProvider>(
     authority_provider: &AP,
     session_id: SessionId,
     session_info: &SessionBoundaryInfo,
+    best_finalized: BlockNumber,
 ) -> Result<CachedData, CacheError> {
-    Ok(match session_id {
-        SessionId(0) => CachedData {
-            session_verifier: authority_provider
-                .authority_data(0)
-                .ok_or(CacheError::UnknownAuthorities(session_id))?
-                .into(),
-            aura_authorities: authority_provider
-                .aura_authorities(0)
-                .ok_or(CacheError::UnknownAuraAuthorities(session_id))?
-                .into_iter()
-                .map(|auth| (None, auth))
-                .collect(),
-        },
+    let (session_verifier, aura_authorities, authority_accounts) = match session_id {
+        SessionId(0) => (
+            authority_provider.authority_data(0),
+            authority_provider.aura_authorities(0),
+            None,
+        ),
         SessionId(id) => {
-            let prev_first = session_info.first_block_of_session(SessionId(id - 1));
-            CachedData {
-                session_verifier: authority_provider
-                    .next_authority_data(prev_first)
-                    .ok_or(CacheError::UnknownAuthorities(session_id))?
-                    .into(),
-                aura_authorities: authority_provider
-                    .next_aura_authorities(prev_first)
-                    .ok_or(CacheError::UnknownAuraAuthorities(session_id))?
-                    .into_iter()
-                    .map(|(acc, auth)| (Some(acc), auth))
-                    .collect(),
-            }
+            let last_block_of_previous_session =
+                session_info.last_block_of_session(SessionId(id - 1));
+            let best_finalized_within_bounds = min(last_block_of_previous_session, best_finalized);
```

Comment:
When is it possible to have `best_finalized` < `last_block_of_previous_session`?

Reply:
Changed.